### PR TITLE
src: implement AsyncPromise class

### PR DESCRIPTION
**Pull request by arc6011**
I created a class to meet a need I have for an asynchronous task which fulfills a promise. Perhaps this would be of some help to others.

### CODE REVIEW

**Comment by arc6011** [comment]
I suppose that if napi_throw fails, the app is probably already going down in flames, eh?
Done.

File: napi-inl.h [hunk 1]
```diff
@@ -3629,6 +3629,293 @@ inline void AsyncWorker::OnWorkComplete(
...
   delete self;
 }
+////////////////////////////////////////////////////////////////////////////////
+// AsyncPromise class
+////////////////////////////////////////////////////////////////////////////////
+
+template <class T>
+inline AsyncPromise<T>::BaseTask::~BaseTask() {
+  if (_work != nullptr) {
+    napi_delete_async_work(_env, _work);
```
[the remark this answers is not in the document]
Comment (arc6011) [reply napi-inl.h]: I suppose that if napi_throw fails, the app is probably already going down in flames, eh?
Done.

**Comment by arc6011** [comment]
Replacing public constructors with static factory methods to make the class signature more consistent with other classes in the API.
I'll commit it as soon as I finish updating the docs and tests.

File: test/asyncpromise.cc [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    Object resource = info[1].As<Object>();
+    std::string data = info[2].As<String>();
+
+    AsyncPromise<TestPromise> async_promise(info.Env(), "TestResource", resource);
+
+    async_promise.Task()->_data = data;
+
+    switch (test)
+    {
+    case Test::succeed_before_queue:
```
Comment (mhdawson) [inline test/asyncpromise.cc]: This should be indented I think to line up with the other cases.

**Comment by mhdawson** [comment]
Overall looks good to me.  Since there is a reasonable amount of code I would like to take another look next week and possibly discuss with the rest of the team in terms of fit with what we have in the API.  Thanks for the contribution!

File: test/asyncpromise.cc (hunk 2 above)
Comment (arc6011) [reply test/asyncpromise.cc]: Thanks, Michael. Now fixed.

**Comment by arc6011** [comment]
AsyncPromise is no longer a template class. The factory methods are now the only templates.

**Comment by DaAitch** [comment]
@arc6011 this is quite a huge PR :)

I like the idea of combining `AsyncWorker` + `Promise` which is a good match.
Nevertheless I always prefer a composition of features over implementing feature specializations, which means I think an implementation of `AsyncWorker` simply using a `Promise` and implement a getter for that promise is much easier to understand then to understand what specific problem an `AsyncPromise` may solve.
My gut feeling tells me, that this can be a new npm module, but is too specific for `node-addon-api`.

**Comment by mhdawson** [comment]
Create this issue to discuss if it should be in scope or not: https://github.com/nodejs/abi-stable-node/issues/363

**Comment by extmchristensen** [comment]
I would expect support for async promises to be  a natual part of this node-addon-api. The feature would be really great to have available soon.

**Comment by KevinEady** [comment]
@mhdawson ,

Going back to the [contributing sources doc](https://github.com/nodejs/node-addon-api/blob/master/CONTRIBUTING.md#source-changes) ... this `AsyncPromise` is neither a NAPI nor NAN functionality, so I feel like this PR doesn't belong in node-addon-api?

This PR is also almost a year old 😄 If it _doesn't_ belong here, should we close it?

**Comment by mhdawson** [comment]
@KevinEady thanks for the ping, I agree we should close it.  Based on our guidance and discussions we'd see this being implemented in a separate module as opposed to being part of node-addon-api. 
 We've chosen to limit node-add-api to be a wrapper of n-api functions with the exception that functionality in nan can also be included in order to support migration.

**Pull request closed without merging**